plugins/kvscheduler/validation.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2020 Pantheon.tech
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kvscheduler
+
+import (
+	"github.com/go-errors/errors"
+	"github.com/golang/protobuf/proto"
+	"go.ligato.io/vpp-agent/v3/pkg/models"
+	"go.ligato.io/vpp-agent/v3/plugins/kvscheduler/api"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+// ValidateSemantically validates given proto messages according to semantic validation(KVDescriptor.Validate)
+// from registered KVDescriptors. If all messages are valid, nil is returned. If all message could be
+// validated, kvscheduler.MessageValidationErrors is returned. In any other case, error is returned.
+//
+// Usage of dynamic proto messages (dynamicpb.Message) described by remotely known models is not supported.
+// The reason for this is that the KVDescriptors can validate only statically generated proto messages and
+// remotely retrieved dynamic proto messages can't be converted to such proto messages (there are
+// no locally available statically generated proto models).
+func (s *Scheduler) ValidateSemantically(messages []proto.Message) error {
+	s.txnLock.Lock()
+	defer s.txnLock.Unlock()
+
+	invalidMessageErrors := make([]*api.InvalidMessageError, 0)
+	for _, message := range messages {
+		originalMessage := message
+
+		// if needed, convert dynamic proto message to statically generated proto message
+		// (validators in descriptors can validate only statically generated proto messages)
+		if dynamicMessage, isDyn := message.(*dynamicpb.Message); isDyn {
+			model, err := models.GetModelFor(message)
+			if err != nil {
+				return errors.Errorf("can't get model for message due to: %v (message=%v)", err, message)
+			}
+			goType := model.LocalGoType() // only for locally known models will return meaningful go type
+			if goType == nil {
+				return errors.Errorf("dynamic messages for remote models are not supported due to "+
```

Comment:
Could there be an option to skip over items with remotely known models instead of returning error?
(i.e. validate everything that can be validated)

Reply:
I forgot here about the remote models that are in local registry (the proxied remote models). It will validate everything that can be validated. See [`695d734`](https://github.com/ligato/vpp-agent/pull/1786/commits/695d73462eb3acdba5360595e11c51f78ac7153f).